File: backend/app/models/dog.py
from sqlalchemy import Column, Integer, String, ForeignKey, Date, Table, Boolean
from sqlalchemy.orm import relationship, backref
from app.core.database import Base
import enum
from sqlalchemy import Enum


class GenderEnum(enum.Enum):
    male = "Male"
    female = "Female"


class StatusEnum(enum.Enum):
    available = "Available"
    sold = "Sold"
    stud = "Available For Stud"
    retired = "Retired"
    active = "Active"
    abkc_champion = "ABKC Champion"


class HealthInfo(Base):
    __tablename__ = "health_info"
    id = Column(Integer, primary_key=True, index=True)
    dog_id = Column(Integer, ForeignKey("dogs.id", ondelete="CASCADE"))
    dna = Column(String(255))
    carrier_status = Column(String(255))
    extra_info = Column(String(255))
    dog = relationship("Dog", back_populates="health_infos")


class Photo(Base):
    __tablename__ = "photos"
    id = Column(Integer, primary_key=True, index=True)
    dog_id = Column(Integer, ForeignKey("dogs.id", ondelete="CASCADE"))
    photo_url = Column(String(255), nullable=False)
    alt = Column(String(255), nullable=False)
    dog = relationship("Dog", back_populates="photos")


dog_production_link = Table(
    "dog_production_link",
    Base.metadata,
    Column("dog_id", Integer, ForeignKey("dogs.id", ondelete="CASCADE")),
    Column("production_id", Integer, ForeignKey("productions.id", ondelete="CASCADE")),
)


class Production(Base):
    __tablename__ = "productions"
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False)

    dob = Column(Date, nullable=True)
    owner = Column(String(255), nullable=True)
    description = Column(String(255), nullable=True)
    sire_id = Column(Integer, ForeignKey("dogs.id"), nullable=True)
    dam_id = Column(Integer, ForeignKey("dogs.id"), nullable=True)
    gender = Column(Enum(GenderEnum), nullable=False)
    profile_photo = Column(String)
    sire = relationship("Dog", foreign_keys=[sire_id], backref="sired_productions")
    dam = relationship("Dog", foreign_keys=[dam_id], backref="damed_productions")
    dogs = relationship(
        "Dog", secondary=dog_production_link, back_populates="productions"
    )


class Dog(Base):
    __tablename__ = "dogs"
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False)
    dob = Column(Date, nullable=True)
    gender = Column(Enum(GenderEnum), nullable=False)
    color = Column(String(255), nullable=True)
    status = Column(Enum(StatusEnum), nullable=True)
    profile_photo = Column(String(255), nullable=True)
    stud_fee = Column(Integer, nullable=True)
    sale_fee = Column(Integer, nullable=True)
    description = Column(String(2500), nullable=True)
    pedigree_link = Column(String(255), nullable=True)
    video_url = Column(String(255), nullable=True)
    parent_male_id = Column(Integer, ForeignKey("dogs.id"), nullable=True)
    parent_female_id = Column(Integer, ForeignKey("dogs.id"), nullable=True)
    is_production = Column(Boolean, default=False)
    kennel_own = Column(Boolean, default=True)
    is_retired = Column(Boolean, default=False)

    health_infos = relationship("HealthInfo", back_populates="dog")
    photos = relationship("Photo", back_populates="dog")
    children = relationship(
        "Dog",
        foreign_keys=[parent_male_id, parent_female_id],
        primaryjoin="or_(Dog.id==Dog.parent_male_id, Dog.id==Dog.parent_female_id)",
        backref=backref("parents", remote_side=[id]),
    )
    productions = relationship(
        "Production", secondary=dog_production_link, back_populates="dogs"
    )

File: frontend/src/api/types/dog.ts
import { CSSProperties } from 'react';
import { GenderEnum, StatusEnum } from './core';

// Main Dog Interface
export interface Dog {
    id: number;
    name: string;
    dob: string;
    gender: GenderEnum;
    color?: string;
    status: StatusEnum;
    profilePhoto: string;
    studFee?: number;
    saleFee?: number;
    description?: string;
    pedigreeLink?: string;
    videoUrl?: string;
    parentMaleId?: number;
    parentFemaleId?: number;
    isRetired: boolean;
    isProduction: boolean;
    healthInfos: HealthInfo[];
    photos: Photo[];
    productions: Production[];
    children: Child[];
    parents: Parent[];
}

export interface Description {
    content: string;
    style?: CSSProperties;
}

// Interfaces for Creating and Updating Dogs
export interface DogCreate {
    name: string;
    dob: string;
    gender: GenderEnum;
    color?: string;
    status?: StatusEnum;
    profilePhoto: string;
    studFee?: number;
    saleFee?: number;
    description?: string;
    pedigreeLink?: string;
    videoUrl?: string;
    parentMaleId?: number;
    parentFemaleId?: number;
    healthInfos?: HealthInfoCreate[];
    galleryPhotos?: string[];
}

export interface DogUpdate {
    name?: string;
    dob?: string;
    gender?: GenderEnum;
    color?: string;
    status?: StatusEnum;
    profilePhoto?: string;
    studFee?: number;
    saleFee?: number;
    description?: string;
    pedigreeLink?: string;
    videoUrl?: string;
    parentMaleId?: number;
    parentFemaleId?: number;
    healthInfos?: HealthInfoCreate[];
    galleryPhotos?: string[];

}

export interface PuppyCreate extends DogCreate {
}

// Health Information Interfaces
export interface HealthInfo {
    id: number;
    dogId: number;
    dna?: string;
    carrierStatus?: string;
    extraInfo?: string;
}

export interface HealthInfoCreate {
    dna?: string;
    carrierStatus?: string;
    extraInfo?: string;
}

// Photo Interfaces
export interface Photo {
    id: number;
    dogId: number;
    photoUrl: string;
    alt: string;
}

export interface PhotoCreate {
    photoUrl: string;
    alt: string;
}

// Parent and Child Interfaces
export interface Parent {
    id: number;
    name: string;
    dob: string;
    gender: GenderEnum;
    profilePhoto: string;
}

export interface Child {
    id: number;
    name: string;
    dob: string;
    gender: GenderEnum;
    profilePhoto: string;
}

// Production Interfaces
export interface Production {
    id: number;
    name: string;
    dob?: string;
    description?: string;
    profilePhoto: string;
    owner?: string;
    parentMaleId?: number;
    parentFemaleId?: number;
    gender: GenderEnum;
}

export interface ProductionCreate {
    name: string;
    dob?: string;
    profilePhoto: string;
    owner?: string;
    parentMaleId?: number;
    parentFemaleId?: number;
    gender: GenderEnum;
}

export interface ProductionUpdate {
    name?: string;
    dob?: string;
    profilePhoto?: string;
    owner?: string;
    parentMaleId?: number;
    parentFemaleId?: number;
    gender?: GenderEnum;
}

export interface ProductionSchema extends Production {
    dogs: Dog[];
}

// Filter interfaces
export interface FilterProps {
    onGenderChange?: (gender?: GenderEnum) => void;
    onStatusChange?: (status: StatusEnum[]) => void;
    onSireChange?: (sire?: Dog) => void;
    onDamChange?: (dam?: Dog) => void;
    onColorChange?: (color: string) => void;
    onRetiredChange?: (retired: boolean) => void;
    gender?: GenderEnum;
    status?: StatusEnum[];
    sire?: { id: number };
    dam?: { id: number };
    color?: string;
    retired?: boolean;
    isGenderDisabled?: boolean;
    isStatusDisabled?: boolean;
    isSireDisabled?: boolean;
    isDamDisabled?: boolean;
    isColorDisabled?: boolean;
}

export interface SelectedFilters {
    gender?: GenderEnum;
    status?: StatusEnum[];
    sire?: { id: number };
    dam?: { id: number };
    color?: string;
    owned?: boolean;
    retired?: boolean;
}
